ggml-cuda.cu

```diff
@@ -5987,6 +6035,527 @@ static  __global__ void im2col_f32_f16(
+                        }
+                        nvcuda::wmma::store_matrix_sync(ss + 16*j*T + 16*cc, mqk[j], T, nvcuda::wmma::mem_col_major);
+                    }
+                }
+            }
+
+            // used to detect blocks full of -INF
+            float smax = -INFINITY;
+
+            // online softmax
+            if (C == 32) {
+                for (int64_t j = 0; j < Q; ++j) {
+                    const int64_t p = lane_id;
+
+                    const float m = M[j];
+                    const float s = __half2float(ss[j*T + p]);
+
+                    smax = warp_reduce_max(max(smax, s));
+                    M[j] = warp_reduce_max(max(M[j], s));
+
+                    const float ms = m == -INFINITY ? 0.0f : expf(m - M[j]);
+                    const float vs = s == -INFINITY ? 0.0f : expf(s - M[j]);
+
+                    S[j] = S[j]*ms + warp_reduce_sum(vs);
+
+                    // create a QxQ diagonal matrix for rescaling the output
+                    if (p == j) {
+                        ss[j*T + C + j] = __float2half(ms);
+                    }
+
+                    // the P matrix from the paper (Q rows, C columns)
+                    ss[j*T + p] = __float2half(vs);
+                }
+            } else {
+                for (int64_t j = 0; j < Q; ++j) {
+                    const float m = M[j];
+
+                    for (int64_t p = lane_id; p < C; p += NW) {
+                        const float s = __half2float(ss[j*T + p]);
+
+                        smax = warp_reduce_max(max(smax, s));
+                        M[j] = warp_reduce_max(max(M[j], s));
+                    }
+
+                    const float ms = m == -INFINITY ? 0.0f : expf(m - M[j]);
+
+                    S[j] = S[j]*ms;
+
+                    // create a QxQ diagonal matrix for rescaling the output
+                    if (lane_id == j) {
+                        ss[j*T + C + j] = ms;
+                    }
+
+                    for (int64_t p = lane_id; p < C; p += NW) {
+                        const float s = ss[j*T + p];
+
+                        const float vs = s == -INFINITY ? 0.0f : expf(s - M[j]);
+
+                        S[j] = S[j] + warp_reduce_sum(vs);
+
+                        // the P matrix from the paper (Q rows, C columns)
+                        ss[j*T + p] = __float2half(vs);
+                    }
+                }
+            }
+
+            // skip -INF blocks
+            if (smax == -INFINITY) {
+                continue;
+            }
+
+            // O = diag(ms)*O
+            for (int64_t j = 0; j < Q16; ++j) {
+                half16x16_a mm;
+                half16x16_b zro;
+
+                nvcuda::wmma::fill_fragment(zro, 0.0);
+                nvcuda::wmma::load_matrix_sync(mm, ss + 16*j*T + C + 16*j, T);
+
+                for (int64_t i = 0; i < D16; ++i) {
+                    nvcuda::wmma::mma_sync(lo[j][i], mm, zro, lo[j][i]);
```

Comment:
This looks incorrect: `lo = mm*0 + lo` ?

Reply:
It might be wrong, but it's an accumulator, so I can't use it directly as a `matrix_b`. I created the `zro` matrix filled with zeros since that part of the code `lo` is currently filled with zeros.

Reply:
I don't understand what you're trying to say but this line is definitely wrong. As Georgi said, this is a no-op that does not change the values of `lo[j][i]` whatsoever.